django/core/handlers/asgi.py

```diff
@@ -179,15 +180,51 @@ async def handle(self, scope, receive, send):
             body_file.close()
             await self.send_response(error_response, send)
             return
-        # Get the response, using the async mode of BaseHandler.
+        # Try to catch a disconnect while getting response
+        try:
+            response = await self.disconnect_and_get_response(request, receive)
+        except RequestAborted:
+            body_file.close()
+            return
+        # Send the response.
+        await self.send_response(response, send)
+
+    async def listen_for_disconnect(self, receive):
+        """Listens for disconnect from the client."""
+        while True:
+            message = await receive()
+            if message["type"] == "http.disconnect":
+                raise RequestAborted()
+
+    async def disconnect_and_get_response(self, request, receive):
+        """Checks for client disconnect while getting response."""
+        tasks = [
+            asyncio.ensure_future(coro(arg))
+            for coro, arg in zip(
+                (
+                    self.run_get_response,
+                    self.listen_for_disconnect,
+                ),
+                (request, receive),
+            )
+        ]
+        done, pending = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
+        try:
+            response = done.pop().result()
```

Comment:
Does the `pending` task needs to be cancelled? 

There should be a test for a view cleaning up after itself. (This will feed into docs for this... — _"If you have a long running request, you may need to clean up in case the client disconnects before you're finished..."_)